File: src/compiler/typescript_to_rust/generators/ic_object/functions/candid_encode.ts
import { Rust } from '../../../../../types';

export function generateIcObjectFunctionCandidEncode(): Rust {
    return /* rust */ `
        fn _azle_ic_candid_encode(
            _this: &boa_engine::JsValue,
            _aargs: &[boa_engine::JsValue],
            _context: &mut boa_engine::Context
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            let candid_string = _aargs.get(0).unwrap().as_string().unwrap();
            let candid_args: candid::IDLArgs = candid_string.parse().unwrap();
            let candid_encoded: Vec<u8> = candid_args.to_bytes().unwrap();

            Ok(candid_encoded.azle_into_js_value(_context))
        }
    `;
}

File: src/compiler/typescript_to_rust/azle_generate/src/body/ic_object/functions/candid_encode.rs
pub fn generate() -> proc_macro2::TokenStream {
    quote::quote! {
        fn candid_encode(
            _this: &boa_engine::JsValue,
            aargs: &[boa_engine::JsValue],
            context: &mut boa_engine::Context,
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            let candid_string: String = aargs
                .get(0)
                .ok_or_else(|| "An argument for 'candidString' was not provided".to_js_error(None))?
                .clone()
                .try_from_vm_value(&mut *context)?;

            let candid_args: candid::IDLArgs =
                candid_string.parse().map_err(|err: candid::error::Error| {
                    boa_engine::error::JsNativeError::error().with_message(err.to_string().as_str())
                })?;
            let candid_encoded: Vec<u8> =
                candid_args
                    .to_bytes()
                    .map_err(|err: candid::error::Error| {
                        boa_engine::error::JsNativeError::error()
                            .with_message(err.to_string().as_str())
                    })?;

            candid_encoded
                .try_into_vm_value(&mut *context)
                .map_err(|vmc_err| vmc_err.to_js_error(None))
        }
    }
}
